### ISSUE TITLE
Datetime scalar transformation loss of resolution

### ISSUE THREAD

**Issue by colinfang** [bokeh/bokeh#2416]
See [notebook](http://nbviewer.ipython.org/gist/colinfang/62431f9245cf311904e8)

The plot cannot go higher resolution than 1 second if the x axis is a list of `datetime` object.  It seems the datetime is converted to a epoch milli-second time internally and the micro-sceond part of the datetime is thrown away. If I provide data with pre-converted datetime with micro-second part , it works out well.

**Comment by colinfang** [comment]
This is Bokeh 9.0

**Comment by colinfang** [comment]
I believe it is this line in the code which caused the problem:

```{python}
isinstance(obj, (dt.datetime, dt.date)):
            return calendar.timegm(obj.timetuple()) * 1000.
```
https://github.com/bokeh/bokeh/blob/0c78cf646ec95d3c003c714c35a388db4b4ad834/bokeh/protocol.py#L85

**Comment by colinfang** [comment]
Also in [code] here(https://github.com/bokeh/bokeh/blob/0c78cf646ec95d3c003c714c35a388db4b4ad834/bokeh/protocol.py#L52)

```{python}
obj.astype('datetime64[ms]').astype('int64').tolist()
```
We seems to always serialize to milli-second epoch time as an integer. Why is it? Do we have to use integer?

**Comment by bryevdv** [comment]
@colinfang I am not sure offhand. This is something that will require some investigation and experimentation. There may be some limitation somewhere that originally caused us to send ms since epoch, but I can't recall what it might have been. In any case fixing this "for real" is probably going to be a little tricky, because we would need to extend the protocol in a way that was not a breaking change, perhaps an optional time base unit of some sort. Very open to input, ideas, PRs, etc.

**Comment by colinfang** [comment]
I understand there must be very good reasons for ms beyond epoch using integer. 

But don't u think it is more like a bug in the scalar case, that it even throws away the millisecond part, which the protocol certainly supports?

**Comment by hhuuggoo** [comment]
There is no reason for throwing away the micros.  I wrote some of it and the astype stuff which throws away micros was just due to laziness.

**Comment by bryevdv** [comment]
@colinfang would you be interested in working up a PR for this to not throw away the microseconds? BTW closing was an accident, I didn't even realize I had done so, marking it "bug" now.

**Comment by bryevdv** [comment]
Also "short term" milestone but a PR before next Tuesday could probably be accepted into 0.9.1

**Comment by colinfang** [comment]
I saw 

```{python}
elif isinstance(obj, dt.time):
            return (obj.hour*3600 + obj.minute*60 + obj.second)*1000 + obj.microsecond / 1000.
```

In this [line](https://github.com/bokeh/bokeh/blob/0c78cf646ec95d3c003c714c35a388db4b4ad834/bokeh/protocol.py#L92)

Noice `obj.microsecond / 1000.` which is a true division, so it returns a float including the microseconds part.

So now I am not very sure whether or not bokeh protocal actually supports the time in float.

**Comment by damianavila** [comment]
Closed by #2425
